Primeiramente devemos importar os nossos dados coletados, armazenando ele no dataframe Df que será tratado

In [1]:
import pandas as pd

# Carregar o arquivo CSV
file_path = 'C:\\Users\\delve\\OneDrive\\Eu\\GitHub\\PO-233-Aprendizado-de-Maquina\\Coleta de dados\\saida.csv'
df = pd.read_csv(file_path, index_col=0)

# Exibir as primeiras linhas do dataframe para entender sua estrutura
df.head()
df

Date  Sales & Services Revenue  Cost of Revenue  Gross Profit   \
0      2010                     78.52             69.0         95.46   
1      2011                     65.96            62.95          70.6   
2      2012                     44.58            36.34          56.7   
3      2013                       9.2            21.36         -6.35   
4      2014                      6.95              5.3          9.69   
...     ...                       ...              ...           ...   
2529   2019                      -1.6            -8.15          5.87   
2530   2020                      0.98             6.54         -4.51   
2531   2021                     16.49            15.28         17.83   
2532   2022                      3.92             3.42          4.45   
2533   2023                      3.42             1.27          5.72   

     Selling, General & Admin   Research & Development   \
0                        32.97                    33.68   
1                        37.74                    36.31   
2                        32.12                    39.19   
3                         7.87                    32.36   
4                        10.74                    34.99   
...                        ...                      ...   
2529                      1.32  - -                       
2530                     17.05  - -                       
2531                     -1.02  - -                       
2532                      7.69  - -                       
2533                      3.31  - -                       

     Operating Income (Loss)  Non-Operating (Income) Loss  Pretax Income   \
0                      140.08                       -52.45         132.21   
1                       83.79                       167.74          84.49   
2                       63.48                        25.78          63.03   
3                       -11.3                       121.46         -10.06   
4                        7.15                       -15.22           6.64   
...                       ...                          ...            ...   
2529                     8.23                      -677.78         -25.34   
2530                    -15.0                        13.27         -25.71   
2531                    30.45                       -27.84          64.12   
2532                     2.81                        16.71          -0.72   
2533                      7.0                        -1.01           9.39   

     Income (Loss) from Cont Ops   ...  Dividends Paid   Cash From Debt   \
0                          145.67  ...  - -              - -               
1                           84.99  ...  - -              - -               
2                           60.99  ...  - -              - -               
3                          -11.25  ...  324.6            - -               
4                            6.68  ...  5.32             -29.21            
...                           ...  ...              ...              ...   
2529                       -16.08  ...            10.61  -42.41            
2530                       -30.14  ...            10.76  87.63             
2531                        74.23  ...             4.59  137.82            
2532                       -15.87  ...             9.63  -69.2             
2533                        20.53  ...             4.47  - -               

      Cash (Repurchase) of Equity  Cash from Financing Activities   \
0                             92.0                           89.59   
1                            -8.88                           14.88   
2                           -19.98                         -217.59   
3                         -3457.89                          864.61   
4                            98.25                          129.25   
...                            ...                             ...   
2529  -65.9                                                  -64.2   
2530  -70.67                      

Etapas que iremos realizar no pré-processamento:
* Transformação de todos nossos dados em floats
* Eliminação de Colunas desnecessárias, que não agregam para o problema
* Eliminação de Colunas que possuem mais que 30% dos seus dados vazios
* Normalização dos dados, assim como normalização da nossa performance

# Eliminação de colunas desnecessárias

Vamos então primeiro mostrar as colunas que possuem valores ausentes, para realizar o tratamento.


Além disso, vamos tornar todos os valores da nossa tabela em numéricos. Caso não sejam numéricos, estarão classificados como NaN

In [6]:
for col in df.columns:
    if df[col].dtype == 'object':
        df[col] = pd.to_numeric(df[col], errors='coerce')

# Mostrar informações sobre valores ausentes
missing_data = df.isnull().sum()

# Exibir as primeiras linhas após conversão
missing_data
#len(df.columns)

Date                           0
Sales & Services Revenue       1
Cost of Revenue              331
Gross Profit                   3
Selling, General & Admin     232
                            ... 
Capital Expenditures         193
Free Cash Flow                 3
Price/Free Cash Flow           2
Cash Flow to Net Income        2
Performance                    1
Length: 77, dtype: int64

In [7]:
# Definir um limite de 10% de valores ausentes
# limite_ausentes = 0.1 * len(df)
missing_percentage = df.isnull().mean()

# Remover colunas com mais de 10% de valores ausentes
# df = df.dropna(thresh=limite_ausentes, axis=1)
cols_to_drop = missing_percentage[missing_percentage > 0.1].index

# Preencher os valores ausentes restantes com a mediana da coluna
# 
df_cleaned = df.drop(columns=cols_to_drop)


# Verificar o resultado
# df
df = df_cleaned
df = df.fillna(df.median())
df.isnull().sum().sum()  # Verificando se ainda há valores ausentes
df = df.drop(df.columns[0], axis = 1)
df.to_csv('Non_Normalized_Data.csv')
df

Sales & Services Revenue   Gross Profit   Selling, General & Admin   \
0                         78.52          95.46                      32.97   
1                         65.96          70.60                      37.74   
2                         44.58          56.70                      32.12   
3                          9.20          -6.35                       7.87   
4                          6.95           9.69                      10.74   
...                         ...            ...                        ...   
2529                      -1.60           5.87                       1.32   
2530                       0.98          -4.51                      17.05   
2531                      16.49          17.83                      -1.02   
2532                       3.92           4.45                       7.69   
2533                       3.42           5.72                       3.31   

      Non-Operating (Income) Loss   Pretax Income   \
0                           -52.45          132.21   
1                           167.74           84.49   
2                            25.78           63.03   
3                           121.46          -10.06   
4                           -15.22            6.64   
...                            ...             ...   
2529                       -677.78          -25.34   
2530                         13.27          -25.71   
2531                        -27.84           64.12   
2532                         16.71           -0.72   
2533                         -1.01            9.39   

      Income (Loss) from Cont Ops   Basic Weighted Avg Shares   \
0                           145.67                        1.84   
1                            84.99                        1.63   
2                            60.99                        1.14   
3                           -11.25                       -1.01   
4                             6.68                       -6.05   
...                            ...                         ...   
2529                        -16.08                       -4.97   
2530                        -30.14                       -1.31   
2531                         74.23                       -1.66   
2532                        -15.87                       -3.70   
2533                         20.53                       -1.75   

      Basic EPS, GAAP   Diluted Weighted Avg Shares   Diluted EPS, GAAP   ...  \
0               141.23                          1.95              140.97  ...   
1                82.02                          1.29               82.63  ...   
2                59.18                          0.93               59.51  ...   
3               -10.34                         -1.45               -9.95  ...   
4                13.54                         -6.12               13.63  ...   
...                ...                           ...                 ...  ...   
2529            -11.70                         -4.86              -11.79  ...   
2530            -29.21                         -1.92              -28.77  ...   
2531             77.16                         -1.63               77.11  ...   
2532            -12.64                         -3.97              -12.39  ...   
2533             22.67                         -1.72               22.64  ...   

      Cash from Operating Activities   Acq of Fixed & Intang   \
0                               83.04                   74.86   
1                              101.82                  251.34   
2                               35.51                   26.17   
3                                5.53                   -3.47   
4                               11.27                    8.12   
...                               ...                     ...   
2529                            11.82                  -16.24   
2530                            -0.76                  -18.37   
2531                            30.73                   43.75   
2532            

Número de colunas antes do tratamento: 77

Número de colunas após o tratamento: 55

Substituimos os valores faltantes restantes pela média dos valores 

-> **tentar mudar para substituir pela media dos valores da mesma categoria**

# Eliminação de Outliers

* Antes de normalizar meus valores, vamos remover os outliers, para isso, primeiro vamos calcular o z-score de cada elemento, isso é, o valor de cada elemento caso a coluna fosse normalizada entre -1 e 1. Com o z-score calculado, vamos escolher um limite para o qual escolheremos os valores nessa nosva distruibuição entre -1 e 1


In [8]:
import numpy as np
data = df
# Calcular z-scores
data_zscores = (data - data.mean()) / data.std()

# Definir limiar de z-score
zscore_threshold = 5

# Filtrar dados
clean_data_zscore = data[(np.abs(data_zscores) < zscore_threshold).all(axis=1)]

# Exibir as primeiras linhas dos dados limpos
data = clean_data_zscore
len(data)
data

Sales & Services Revenue   Gross Profit   Selling, General & Admin   \
0                         78.52          95.46                      32.97   
1                         65.96          70.60                      37.74   
2                         44.58          56.70                      32.12   
3                          9.20          -6.35                       7.87   
4                          6.95           9.69                      10.74   
...                         ...            ...                        ...   
2529                      -1.60           5.87                       1.32   
2530                       0.98          -4.51                      17.05   
2531                      16.49          17.83                      -1.02   
2532                       3.92           4.45                       7.69   
2533                       3.42           5.72                       3.31   

      Non-Operating (Income) Loss   Pretax Income   \
0                           -52.45          132.21   
1                           167.74           84.49   
2                            25.78           63.03   
3                           121.46          -10.06   
4                           -15.22            6.64   
...                            ...             ...   
2529                       -677.78          -25.34   
2530                         13.27          -25.71   
2531                        -27.84           64.12   
2532                         16.71           -0.72   
2533                         -1.01            9.39   

      Income (Loss) from Cont Ops   Basic Weighted Avg Shares   \
0                           145.67                        1.84   
1                            84.99                        1.63   
2                            60.99                        1.14   
3                           -11.25                       -1.01   
4                             6.68                       -6.05   
...                            ...                         ...   
2529                        -16.08                       -4.97   
2530                        -30.14                       -1.31   
2531                         74.23                       -1.66   
2532                        -15.87                       -3.70   
2533                         20.53                       -1.75   

      Basic EPS, GAAP   Diluted Weighted Avg Shares   Diluted EPS, GAAP   ...  \
0               141.23                          1.95              140.97  ...   
1                82.02                          1.29               82.63  ...   
2                59.18                          0.93               59.51  ...   
3               -10.34                         -1.45               -9.95  ...   
4                13.54                         -6.12               13.63  ...   
...                ...                           ...                 ...  ...   
2529            -11.70                         -4.86              -11.79  ...   
2530            -29.21                         -1.92              -28.77  ...   
2531             77.16                         -1.63               77.11  ...   
2532            -12.64                         -3.97              -12.39  ...   
2533             22.67                         -1.72               22.64  ...   

      Cash from Operating Activities   Acq of Fixed & Intang   \
0                               83.04                   74.86   
1                              101.82                  251.34   
2                               35.51                   26.17   
3                                5.53                   -3.47   
4                               11.27                    8.12   
...                               ...                     ...   
2529                            11.82                  -16.24   
2530                            -0.76                  -18.37   
2531                            30.73                   43.75   
2532            

* Antes da eliminação, estavamos com 2533 linhas, após ficamos com apenas 2272

# Normalização
Agora o próximo passo é normalizar minhas colunas

valor = valor / std

Escolhi essa forma de normalizar para mantes valores positivos e valores negativos sem perder a distribuição.

In [9]:
from sklearn.preprocessing import MinMaxScaler

means = data.mean()
stds = data.std()
normalized_data = ((data) / stds)

# Exibir as primeiras linhas dos dados normalizados
data = normalized_data
data.head()

Sales & Services Revenue   Gross Profit   Selling, General & Admin   \
0                   5.184228       3.270406                   1.836533   
1                   4.354963       2.418717                   2.102237   
2                   2.943363       1.942510                   1.789186   
3                   0.607424      -0.217547                   0.438384   
4                   0.458869       0.331974                   0.598252   

   Non-Operating (Income) Loss   Pretax Income   Income (Loss) from Cont Ops   \
0                     -0.129995        0.374486                      0.684311   
1                      0.415736        0.239319                      0.399256   
2                      0.063895        0.178533                      0.286511   
3                      0.301033       -0.028495                     -0.052849   
4                     -0.037722        0.018808                      0.031381   

   Basic Weighted Avg Shares   Basic EPS, GAAP   Diluted Weighted Avg Shares   \
0                    0.281492          0.694453                      0.294091   
1                    0.249365          0.403307                      0.194553   
2                    0.174403          0.290998                      0.140259   
3                   -0.154515         -0.050844                     -0.218683   
4                   -0.925558          0.066579                     -0.922993   

   Diluted EPS, GAAP   ...  Cash from Operating Activities   \
0            0.696380  ...                         0.649951   
1            0.408185  ...                         0.796941   
2            0.293974  ...                         0.277935   
3           -0.049152  ...                         0.043283   
4            0.067331  ...                         0.088210   

   Acq of Fixed & Intang   Cash from Investing Activities   \
0                0.702489                        -0.011974   
1                2.358583                         0.111835   
2                0.245580                         0.011268   
3               -0.032563                        -0.017480   
4                0.076198                        -0.019334   

   Cash from Financing Activities   Net Changes in Cash   \
0                         0.111563             -0.053267   
1                         0.018530             -0.034665   
2                        -0.270957             -0.045913   
3                         1.076669              0.077464   
4                         0.160951             -0.031230   

   Capital Expenditures   Free Cash Flow   Price/Free Cash Flow   \
0               0.702489         0.281841              -0.036114   
1               2.358583         0.276550              -0.103958   
2               0.245580         0.126703               0.088897   
3              -0.032563         0.025354              -0.092919   
4               0.076198         0.039890               0.070369   

   Cash Flow to Net Income   Performance  
0                 -0.077416     1.678843  
1                  0.027638     1.065240  
2                 -0.048077     0.798523  
3                  0.057431    -0.897314  
4                  0.013060     1.218211  

[5 rows x 55 columns]

In [10]:
data.to_csv('Normalized_Data.csv')
data

Sales & Services Revenue   Gross Profit   Selling, General & Admin   \
0                      5.184228       3.270406                   1.836533   
1                      4.354963       2.418717                   2.102237   
2                      2.943363       1.942510                   1.789186   
3                      0.607424      -0.217547                   0.438384   
4                      0.458869       0.331974                   0.598252   
...                         ...            ...                        ...   
2529                  -0.105639       0.201103                   0.073528   
2530                   0.064704      -0.154510                   0.949739   
2531                   1.088741       0.610846                  -0.056817   
2532                   0.258815       0.152455                   0.428357   
2533                   0.225803       0.195964                   0.184377   

      Non-Operating (Income) Loss   Pretax Income   \
0                        -0.129995        0.374486   
1                         0.415736        0.239319   
2                         0.063895        0.178533   
3                         0.301033       -0.028495   
4                        -0.037722        0.018808   
...                            ...             ...   
2529                     -1.679849       -0.071776   
2530                      0.032889       -0.072824   
2531                     -0.069000        0.181621   
2532                      0.041415       -0.002039   
2533                     -0.002503        0.026597   

      Income (Loss) from Cont Ops   Basic Weighted Avg Shares   \
0                         0.684311                    0.281492   
1                         0.399256                    0.249365   
2                         0.286511                    0.174403   
3                        -0.052849                   -0.154515   
4                         0.031381                   -0.925558   
...                            ...                         ...   
2529                     -0.075539                   -0.760335   
2530                     -0.141588                   -0.200410   
2531                      0.348709                   -0.253955   
2532                     -0.074552                   -0.566044   
2533                      0.096443                   -0.267723   

      Basic EPS, GAAP   Diluted Weighted Avg Shares   Diluted EPS, GAAP   ...  \
0             0.694453                      0.294091            0.696380  ...   
1             0.403307                      0.194553            0.408185  ...   
2             0.290998                      0.140259            0.293974  ...   
3            -0.050844                     -0.218683           -0.049152  ...   
4             0.066579                     -0.922993            0.067331  ...   
...                ...                           ...                 ...  ...   
2529         -0.057531                     -0.732965           -0.058242  ...   
2530         -0.143631                     -0.289567           -0.142121  ...   
2531          0.379409                     -0.245830            0.380917  ...   
2532         -0.062153                     -0.598739           -0.061206  ...   
2533          0.111472                     -0.259403            0.111840  ...   

      Cash from Operating Activities   Acq of Fixed & Intang   \
0                            0.649951                0.702489   
1                            0.796941                2.358583   
2                            0.277935                0.245580   
3                            0.043283               -0.032563   
4                            0.088210                0.076198   
...                               ...                     ...   
2529                         0.092515               -0.152397   
2530                        -0.005948               -0.172385   
2531                         0.240523                0.410551   
2532            